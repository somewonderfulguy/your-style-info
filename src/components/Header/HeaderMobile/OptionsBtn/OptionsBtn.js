import React, {useRef} from 'react'
import {bool, func} from 'prop-types'
import {useSpring, animated} from 'react-spring'

import {useTheme} from 'helpers/contexts'
import styles from './OptionsBtn.module.css'

const propTypes = {
  isOpen: bool,
  onClick: func
}

const defaultProps = {
  isOpen: false,
  onClick: () => {}
}

const OptionsBtn = ({isOpen, onClick}) => {
  const optionsBtnDOM = useRef(null)
  const {isDarkTheme} = useTheme()
  const optionsHeight = optionsBtnDOM.current ? optionsBtnDOM.current.clientHeight : 0

  // TODO delete this, awful idea
  const dotSize = +(
    getComputedStyle(document.documentElement).getPropertyValue('--mobile-menu-options-dot-diameter')
      .replace('px', '')
      || 4
  )
  const shrinkedDotSize = dotSize / 2

  const middlePosition = (optionsHeight / 2) - shrinkedDotSize / 2

  const {timeline} = useSpring({
    config: {duration: 300},
    timeline: isOpen ? 1 : 0
  })

  const range = [0, 0.4, 0.6, 1]

  const topDot = {
    top: timeline.interpolate(range, [0, middlePosition, middlePosition, middlePosition]),
    width: timeline.interpolate(range, [4, 2, 2, 24]),
    height: timeline.interpolate(range, [4, 2, 2, 2]),
    borderRadius: timeline.interpolate(range, ['50%', '0', '0', '0'])
  }

  const bottomDot = {
    bottom: topDot.top,
    width: topDot.width,
    height: topDot.height,
    borderRadius: timeline.interpolate(range, ['50%', '0', '0', '0'])
  }

  const middleDot = {
    opacity: timeline.interpolate({range, output: [1, 0, 0, 0]})
  }

  return (
    <button ref={optionsBtnDOM} onClick={onClick} className={styles[isDarkTheme ? 'darkBtn' : 'btn']} type="button">
      <animated.div className={styles.dot} style={topDot} />
      <animated.div className={styles.dot} style={middleDot} />
      <animated.div className={styles.dot} style={bottomDot} />
    </button>
  )
}

OptionsBtn.propTypes = propTypes
OptionsBtn.defaultProps = defaultProps

export default OptionsBtn