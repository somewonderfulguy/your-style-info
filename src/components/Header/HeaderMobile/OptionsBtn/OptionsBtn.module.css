:root {
  --mobile-menu-options-size: 20px;
  --mobile-menu-options-dot-diameter: 4px;
}

.btn {
  position: relative;
  width: var(--mobile-menu-options-size);
  height: var(--mobile-menu-options-size);

  display: flex;
  justify-content: center;
  
  outline: none;
}

.darkBtn { composes: btn; }

.dot {
  width: var(--mobile-menu-options-dot-diameter);
  height: var(--mobile-menu-options-dot-diameter);

  background-color: #000;
  border-radius: 50%;

  position: absolute;
  transition: background-color .2s;
}

.darkBtn .dot {
  background-color: lightgray;
}

.dot:nth-child(1) {top: 0; transform: rotate(-45deg);}
.dot:nth-child(2) {top: calc(50% - var(--mobile-menu-options-dot-diameter) / 2);}
.dot:nth-child(3) {bottom: 0; transform: rotate(45deg);}