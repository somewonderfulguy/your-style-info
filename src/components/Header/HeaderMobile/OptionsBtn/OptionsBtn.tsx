import React, { forwardRef, useImperativeHandle, useRef } from 'react'
import { useSpring, animated } from 'react-spring'

import { useThemeState } from 'contexts'
import styles from './OptionsBtn.module.css'

type propType = {
  isOpen?: boolean
  onClick?: () => void
}

const OptionsBtn = forwardRef(
  ({ isOpen = false, onClick = () => {} }: propType, ref) => {
    const optionsBtnDOM = useRef<HTMLButtonElement>(null)
    useImperativeHandle(ref, () => ({ optionsBtnDOM }))

    const isDarkTheme = useThemeState()

    const optionsHeight = optionsBtnDOM.current
      ? optionsBtnDOM.current.clientHeight
      : 0

    const dotSize = 4
    const shrinkedDotSize = dotSize / 2
    const middlePosition = optionsHeight / 2 - shrinkedDotSize / 2

    const { timeline: t } = useSpring({
      config: { duration: 300 },
      timeline: isOpen ? 1 : 0
    })
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const timeline = t as any

    const range = [0, 0.4, 0.6, 1]

    const topDot = {
      top: timeline.interpolate(range, [
        0,
        middlePosition,
        middlePosition,
        middlePosition
      ]),
      width: timeline.interpolate(range, [
        dotSize,
        shrinkedDotSize,
        shrinkedDotSize,
        24
      ]),
      height: timeline.interpolate(range, [
        dotSize,
        shrinkedDotSize,
        shrinkedDotSize,
        2
      ]),
      borderRadius: timeline.interpolate(range, ['50%', '0', '0', '0'])
    }

    const bottomDot = {
      bottom: topDot.top,
      width: topDot.width,
      height: topDot.height,
      borderRadius: timeline.interpolate(range, ['50%', '0', '0', '0'])
    }

    const middleDot = {
      width: dotSize,
      height: dotSize,
      opacity: timeline.interpolate({ range, output: [1, 0, 0, 0] })
    }

    return (
      <button
        ref={optionsBtnDOM}
        onClick={onClick}
        className={styles[isDarkTheme ? 'darkBtn' : 'btn']}
        type="button"
        // TODO: do it dynamic
        title="Options"
      >
        <animated.div className={styles.dot} style={topDot} />
        <animated.div className={styles.dot} style={middleDot} />
        <animated.div className={styles.dot} style={bottomDot} />
      </button>
    )
  }
)
OptionsBtn.displayName = 'OptionsBtn'

export default OptionsBtn
