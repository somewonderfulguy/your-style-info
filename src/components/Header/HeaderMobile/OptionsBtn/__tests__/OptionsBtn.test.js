import React from 'react'
import {render} from '@testing-library/react'

import OptionsBtn from '../'

test('should match snapshot', () => {
  const {asFragment} = render(<OptionsBtn />)
  expect(asFragment()).toMatchSnapshot()
})