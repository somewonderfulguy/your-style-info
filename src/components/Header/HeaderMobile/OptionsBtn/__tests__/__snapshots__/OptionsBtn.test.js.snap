// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
    type="button"
  >
    <div
      style="top: 0px; width: 4px; height: 4px; border-radius: 50%;"
    />
    <div
      style="opacity: 1;"
    />
    <div
      style="bottom: 0px; width: 4px; height: 4px; border-radius: 50%;"
    />
  </button>
</DocumentFragment>
`;
